### PR TITLE
MAINT: Python3 improvements to refguide_check.py 

### PR DESCRIPTION
**Pull request by sethtroisi**
This builds on https://github.com/scipy/scipy/pull/11433 (now submitted) (which has a little personal intro)

This is 3 changes I submitted to numpy (see https://github.com/numpy/numpy/issues/15306#issuecomment-579428830 for numpy commits)

```
MAINT: Use contextmanager in _run_doctests
DOC: Update refguide_check note on how to skip code #15449
MAINT: Cleanup references to python2
```
I've additionally sorted the imports to be, what I believe is, PEP8 compliant.
```
MAINT: reorder tools/refguide_check imports
```

If you would prefer this as 4 git commits instead of the squashed
`DEV: upstream changes to refguide_check from numpy`
I can try to unbundle them

### CODE REVIEW

**Review by andyfaff — COMMENTED** [review]
I'll leave it to others to comment on the refguide testing, but you mistakenly included changes from a previous commit, that are also present in another PR

**Comment by sethtroisi** [comment]
@andyfaff I kept the other commit included so that tests would pass in this PR. I'll rebase when/if the other PR get's pulled.

File: tools/refguide_check.py [hunk 1]
```diff
@@ -24,21 +24,23 @@
...
-from doctest import NORMALIZE_WHITESPACE, ELLIPSIS, IGNORE_EXCEPTION_DETAIL
+import sys
+import tempfile
+import warnings
+import docutils.core
 from argparse import ArgumentParser
+from contextlib import contextmanager, redirect_stderr
+from doctest import NORMALIZE_WHITESPACE, ELLIPSIS, IGNORE_EXCEPTION_DETAIL
+
+from docutils.parsers.rst import directives
```
Comment (pvanmulbregt) [inline tools/refguide_check.py]: Why the change in the import ordering? 
The reordering is not called out in the commit or PR description, so asking about the rationale for it.  [I wondered if it might have been making the imports PEP8-compliant but that doesn't seem to have been the case.]

Comment (sethtroisi) [reply tools/refguide_check.py]: I was trying to make it closer to PEP8-compliant, I've reordered them to be what I believe is PEP8 compliant (I'm not sure how to sort the `from numpydoc.docscrape_sphinx import` which requires modifying sys.path) and to lead comment in this PR.

Comment (pvanmulbregt) [reply tools/refguide_check.py]: Given that the `from numpydoc.docscrape_sphinx import ...` has to occur after the `sys.path` change, there is really no other place it can go.

**Comment by pvanmulbregt** [comment]
What testing is in place for `tools/refguide_check.py` and the other `tools/` scripts?  What criteria can be used to indicate a PR inadvertently introduced an different behavior?

**Comment by rgommers** [comment]
> What testing is in place for `tools/refguide_check.py` and the other `tools/` scripts? What criteria can be used to indicate a PR inadvertently introduced an different behavior?

Not much, only that CI still passes in the refguide check job. If it would silently detect fewer tests, that is hard to see though.

Other scripts in `tools/` may have no automated tooling. Most are just used at release time; it's not uncommon to find some last-minute issues but usually they're not hard to fix.

**Comment by tylerjereddy** [comment]
If we had the ability to trace (show coverage for) refguide docstring parsing then a constant coverage number might help with confidence a bit. But, we can't at the moment, and even that wouldn't guarantee no change in verification behavior.

**Comment by ev-br** [comment]
> If we had the ability to trace (show coverage for) refguide docstring parsing then a constant coverage number might help with confidence a bit. But, we can't at the moment, and even that wouldn't guarantee no change in verification behavior.

As a quick hack, one may check `len(tests)` at
https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L673
and/or
https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L717

**Comment by sethtroisi** [comment]
I verified that `python tools/refguide_check.py > before.txt 2> before.stderr` matched before and after.
I also checked printed `len(tests)` in both places and verified they match. 

[Un]Luckily one of the test ("scipy/optimize/optimize.py", line 889, in check_grad) was broken so I also verified that output

**Comment by ev-br** [comment]
Ok, let's get this in. Thanks @sethtroisi , all.

**Comment by sethtroisi** [comment]
Thanks :)

**Comment by mdhaber** [comment]
Thanks @sethtroisi! If you'd like your name added to [THANKS.txt](https://github.com/scipy/scipy/blob/master/THANKS.txt), please make an entry [here](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.5.0*).
